### PR TITLE
Enhancement `dao.js` tidying

### PR DESCRIPTION
**Pull request by kiserd**
- Offloaded some work from the main functions into descriptive helper functions to improve readability.
- Some functions were calling several independent asynchronous functions in a row and requiring that they `await` their neighbors resolve. However, given that they are independent, it was best to have them do their fetching simultaneously and instruct the remainder of the function to `await` their collective resolve. This was a great suggestion by Josh.

### CODE REVIEW

File: portfolio-project-matching-app/backend/dao.js [hunk 1]
```diff
@@ -139,33 +139,8 @@ const createAssociation = async (coll, id1, id2) => {
...
-        return -1;
-    }
-    // handle case where id2 does not exist in coll2
-    else if (!id2Snap.exists()) {
-        console.log(`Invalid '${coll2}' id: '${id2}' does not exist`);
-        return -1;
-    }
-    // handle case where id1_id2 already exists in coll
-    else if (collSnap.exists()) {
-        console.log(`Invalid id1 id2 combo: '${id1}_${id2}' already exists in '${coll}'`);
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:168]: This was just clogging up the main function and hindering readability.

File: portfolio-project-matching-app/backend/dao.js [hunk 2]
```diff
@@ -176,20 +151,8 @@ const createAssociation = async (coll, id1, id2) => {
...
-            // handle case where project needs closed
-            const payloadNeedsUpdated = id1Snap.data().census === id1Snap.data().capacity - 1;
-            if (payloadNeedsUpdated){
-                payload.open = false;
-            }
-            const projectSnapNew = await updateDoc(coll1, id1, payload);
-            console.log(`Updated '${coll1}' document id '${id1}' census to ${projectSnapNew.data().census}`);
-            if (payloadNeedsUpdated) {
-                console.log(`Closed '${coll1}' document id '${id1}'`);
-            }
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:192]: Same here, clogging up the main function and making it impossible to read.

File: portfolio-project-matching-app/backend/dao.js [hunk 3]
```diff
@@ -244,33 +292,78 @@ const readAllDocs = async (coll) => {
...
+    }
+    // handle case of 'users' collection
+    else if (coll === 'users') {
+        return User.fromDocSnapshot(doc.id, doc);
+    }
+    // handle case of 'technologies collection
+    else if (coll === 'technologies') {
+        return Technology.fromDocSnapshot(doc.id, doc);
+    }
+}
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:352]: These ought to come in handy down the road. There are a couple dedicated `getAll___()` that could be consolidated into a single function.

File: portfolio-project-matching-app/backend/dao.js [hunk 4]
```diff
@@ -356,9 +449,11 @@ const deleteAssociation = async (coll, id1, id2) => {
...
     const [coll1, coll2] = coll.split('_');
     // get document snapshots for invalid input handling
-    const id1Snap = await getDocSnapshotById(coll1, id1);
-    const id2Snap = await getDocSnapshotById(coll2, id2);
-    const collSnap = await getDocSnapshotById(coll, `${id1}_${id2}`);
+    const [id1Snap, id2Snap, collSnap] = await Promise.all([
+        getDocSnapshotById(coll1, id1),
+        getDocSnapshotById(coll2, id2),
+        getDocSnapshotById(coll, `${id1}_${id2}`)
+    ]);
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:456]: Great call by Josh on these types of improvements. No need to await each other's resolve.. let them do their work and move forward when they all finish asynchronously.

File: portfolio-project-matching-app/pages/test_dao.js [hunk 5]
```diff
@@ -99,7 +99,7 @@ const test_createProject = () => {
         // const userId = 'PdJa7Nq3LJCbEOFxA2Vj';
         // await deleteAssociation('projects_users', projectId, userId); 
 
-        const coll = 'technologies';
+        const coll = 'projects';
```
Comment (kiserd) [inline portfolio-project-matching-app/pages/test_dao.js:102]: Just doing some testing in here, nothing to see.

**Review by ylijokic — APPROVED** [review]

File: portfolio-project-matching-app/backend/dao.js [hunk 6]
```diff
@@ -364,6 +365,25 @@ const collectionIsValid = async (coll) => {
                     is valid
     */
     return coll === 'projects' || coll === 'users' || coll === 'technologies';
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:367]: I Missed a closing bracket when I merged these together. What an idiot.

**Pull request merged**